backend/wallet/services/offchain/payment.py

```diff
@@ -0,0 +1,234 @@
+        vasp_address=vasp_address,
+        my_address=generate_my_address(account_id),
+        reference_id=reference_id,
+        merchant_name=merchant_name,
+        action=action,
+        currency=currency,
+        amount=amount,
+        expiration=datetime.fromtimestamp(expiration) if expiration else None,
+    )
+
+    save_payment(payment_command)
+
+
+def approve_payment(
+    account_id: int, reference_id: str, init_required: Optional[bool] = True
+):
+    payment_model = storage.get_payment_details(reference_id)
+
+    if not payment_model:
+        raise PaymentNotFoundError(f"Could not find payment {reference_id}")
+
+    if payment_model.action == "charge":
+        if init_required:
+            send_init_charge_payment_request(payment_model, account_id)
+    elif payment_model.action == "auth":
+        if init_required:
+            send_init_auth_payment_request(payment_model, account_id)
+
+        lock_funds(account_id, payment_model, reference_id)
+    else:
+        raise P2MGeneralError(f"Unsupported action type {payment_model.action}")
+
+
+def lock_funds(account_id, payment_model, reference_id):
+    my_address, my_sub_address = utils.account_address_and_subaddress(
+        payment_model.my_address
+    )
+    vasp_address, vasp_sub_address = utils.account_address_and_subaddress(
+        payment_model.vasp_address
+    )
+    storage.add_transaction(
+        amount=payment_model.amount,
+        currency=DiemCurrency[payment_model.currency],
+        payment_type=TransactionType.EXTERNAL,
+        status=TransactionStatus.LOCKED,
+        source_id=account_id,
+        source_address=my_address,
+        source_subaddress=my_sub_address,
+        destination_id=None,
+        destination_address=vasp_address,
+        destination_subaddress=vasp_sub_address,
+        reference_id=reference_id,
+    )
+
+
+def send_init_auth_payment_request(payment_model, account_id):
+    user = storage.get_user(account_id)
+
+    try:
+        utils.offchain_client().send_request(
+            request_sender_address=payment_model.my_address,
+            counterparty_account_id=payment_model.vasp_address,
+            request_bytes=jws.serialize(
+                new_init_auth_command(
+                    reference_id=payment_model.reference_id,
+                    vasp_address=payment_model.vasp_address,
+                    sender_name=user.first_name,
+                    sender_sure_name=user.last_name,
+                    sender_city=user.city,
+                    sender_country=user.country,
+                    sender_line1=user.address_1,
+                    sender_line2=user.address_2,
+                    sender_postal_code=user.zip,
+                    sender_state=user.state,
+                    sender_national_id_value="",
+                    sender_national_id_type="",
+                ),
+                utils.compliance_private_key().sign,
+            ),
+        )
+    except Exception as e:
```

Comment:
This is very discouraged. Is it possible to know the specific exception type? And why would you catch it at all if you do nothing with it? Won't the outer layers handle it?

Reply:
I know, but there is too many options can thrown from offchain client so we "catch" generic one, I added handling in the end point which call this method ehich will return 500 in case of P2MGeneralError